File: Aces/Maneuvers.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Aces
{
    public static class Maneuvers
    {
        public const string Stall_left = "Stall left";
        public const string Turn_left = "Turn left";
        public const string Weave_left_right = "Weave left-right";
        public const string Stall = "Stall";
        public const string Stall_right = "Stall right";
        public const string Turn_right = "Turn right";
        public const string Rotary_turn = "Rotary turn";
        public const string Weave_right_left = "Weave right-left";
        public const string Cruise_then_left = "Cruise then left";
        public const string Left_then_cruise = "Left then cruise";
        public const string Wing_left = "Wing left";
        public const string Straight = "Straight";
        public const string Immlemann = "Immleman";
        public const string Slip_left = "Slip left";
        public const string Slip_right = "Slip right";
        public const string Cruise_then_right = "Cruise then right";
        public const string Right_then_cruise = "Right then cruise";
        public const string Wing_right = "Wing right";
        public const string Fast_then_left = "Fast then left";
        public const string Left_then_fast = "Left then fast";
        public const string Fast = "Fast";
        public const string Barrel_roll_left = "Barrel roll left";
        public const string Barrel_roll_right = "Barrel roll right";
        public const string Fast_then_right = "Fast then right";
        public const string Right_then_fast = "Right then fast";

        public class Details
        {
            public Details() { }
            public Details(String direction, Boolean difficult) 
            {
                Difficult = difficult;
                Direction = direction;
            }
            public String Direction { get; set; }
            public Boolean Difficult { get; set; }
            public String Speed { get; set; }
        }

        public static Details GetDetails(String maneuver)
        {
            if (maneuver != null && maneuvers.ContainsKey(maneuver))
                return maneuvers[maneuver];
            else
                return null;
        }

        public static IDictionary<String, Details> All { get { return maneuvers; } }

        public static List<string> Select(Predicate<Maneuvers.Details> pred)
        {
            List<string> result = new List<string>();
            foreach (string maneuver in maneuvers.Keys)
            {
                if (pred(maneuvers[maneuver])) result.Add(maneuver);
            }
            return result;
        }
        private static IDictionary<String, Details> maneuvers = new Dictionary<String, Details>()
        {
            // 0. Slow (no power dive)
            { Stall_left, new Details { Direction = "left", Speed = "slow", Difficult = false } }, 
            { Turn_left, new Details { Direction = "left", Speed = "slow", Difficult = false } }, 
            { Weave_left_right, new Details { Direction = "right", Speed = "slow", Difficult = true } }, 
    
            { Stall, new Details { Direction = "straight", Speed = "slow", Difficult = false } }, 
    
            { Stall_right, new Details { Direction = "right", Speed = "slow", Difficult = false } }, 
            { Turn_right, new Details { Direction = "right", Speed = "slow", Difficult = false } }, 
            { Rotary_turn, new Details { Direction = "right", Speed = "slow", Difficult = false } }, 
            { Weave_right_left, new Details { Direction = "left", Speed = "slow", Difficult = true } }, 
    
            // 1. Cruising
            { Cruise_then_left, new Details { Direction = "left", Speed = "cruise", Difficult = false } }, 
            { Left_then_cruise, new Details { Direction = "left", Speed = "cruise", Difficult = false } }, 
            { Wing_left, new Details { Direction = "left", Speed = "cruise", Difficult = true } }, 
    
            { Straight, new Details { Direction = "straight", Speed = "cruise", Difficult = false } }, 
            { Immlemann, new Details { Direction = "straight", Speed = "cruise", Difficult = true } }, 
            { Slip_left, new Details { Direction = "straight", Speed = "cruise", Difficult = true } }, 
            { Slip_right, new Details { Direction = "straight", Speed = "cruise", Difficult = true } }, 
    
            { Cruise_then_right, new Details { Direction = "right", Speed = "cruise", Difficult = false } }, 
            { Right_then_cruise, new Details { Direction = "right", Speed = "cruise", Difficult = false } }, 
            { Wing_right, new Details { Direction = "right", Speed = "cruise", Difficult = true } }, 
    
            // 2. Fast (no climb)
            { Fast_then_left, new Details { Direction = "left", Speed = "fast", Difficult = false } }, 
            { Left_then_fast, new Details { Direction = "left", Speed = "fast", Difficult = false } }, 
    
            { Fast, new Details { Direction = "straight", Speed = "fast", Difficult = false } }, 
            { Barrel_roll_left, new Details { Direction = "straight", Speed = "fast", Difficult = true } }, 
            { Barrel_roll_right, new Details { Direction = "straight", Speed = "fast", Difficult = true } }, 
    
            { Fast_then_right, new Details { Direction = "right", Speed = "fast", Difficult = false } }, 
            { Right_then_fast, new Details { Direction = "right", Speed = "fast", Difficult = false } }, 
        };
    }
}

File: ace-of-aces-react/src/models/maneuvers.ts
// src/models/maneuvers.ts

import { ManeuverDetails } from './types';

export const ManeuverConstants = {
  Stall_left: "Stall left",
  Turn_left: "Turn left",
  Weave_left_right: "Weave left-right",
  Stall: "Stall",
  Stall_right: "Stall right",
  Turn_right: "Turn right",
  Rotary_turn: "Rotary turn",
  Weave_right_left: "Weave right-left",
  Cruise_then_left: "Cruise then left",
  Left_then_cruise: "Left then cruise",
  Wing_left: "Wing left",
  Straight: "Straight",
  Immlemann: "Immleman",
  Slip_left: "Slip left",
  Slip_right: "Slip right",
  Cruise_then_right: "Cruise then right",
  Right_then_cruise: "Right then cruise",
  Wing_right: "Wing right",
  Fast_then_left: "Fast then left",
  Left_then_fast: "Left then fast",
  Fast: "Fast",
  Barrel_roll_left: "Barrel roll left",
  Barrel_roll_right: "Barrel roll right",
  Fast_then_right: "Fast then right",
  Right_then_fast: "Right then fast",
};

export const ManeuverDetailsMap: { [key: string]: ManeuverDetails } = {
  [ManeuverConstants.Stall_left]: { direction: "left", speed: "slow", difficult: false },
  [ManeuverConstants.Turn_left]: { direction: "left", speed: "slow", difficult: false },
  [ManeuverConstants.Weave_left_right]: { direction: "right", speed: "slow", difficult: true },
  [ManeuverConstants.Stall]: { direction: "straight", speed: "slow", difficult: false },
  [ManeuverConstants.Stall_right]: { direction: "right", speed: "slow", difficult: false },
  [ManeuverConstants.Turn_right]: { direction: "right", speed: "slow", difficult: false },
  [ManeuverConstants.Rotary_turn]: { direction: "right", speed: "slow", difficult: false },
  [ManeuverConstants.Weave_right_left]: { direction: "left", speed: "slow", difficult: true },
  [ManeuverConstants.Cruise_then_left]: { direction: "left", speed: "cruise", difficult: false },
  [ManeuverConstants.Left_then_cruise]: { direction: "left", speed: "cruise", difficult: false },
  [ManeuverConstants.Wing_left]: { direction: "left", speed: "cruise", difficult: true },
  [ManeuverConstants.Straight]: { direction: "straight", speed: "cruise", difficult: false },
  [ManeuverConstants.Immlemann]: { direction: "straight", speed: "cruise", difficult: true },
  [ManeuverConstants.Slip_left]: { direction: "straight", speed: "cruise", difficult: true },
  [ManeuverConstants.Slip_right]: { direction: "straight", speed: "cruise", difficult: true },
  [ManeuverConstants.Cruise_then_right]: { direction: "right", speed: "cruise", difficult: false },
  [ManeuverConstants.Right_then_cruise]: { direction: "right", speed: "cruise", difficult: false },
  [ManeuverConstants.Wing_right]: { direction: "right", speed: "cruise", difficult: true },
  [ManeuverConstants.Fast_then_left]: { direction: "left", speed: "fast", difficult: false },
  [ManeuverConstants.Left_then_fast]: { direction: "left", speed: "fast", difficult: false },
  [ManeuverConstants.Fast]: { direction: "straight", speed: "fast", difficult: false },
  [ManeuverConstants.Barrel_roll_left]: { direction: "straight", speed: "fast", difficult: true },
  [ManeuverConstants.Barrel_roll_right]: { direction: "straight", speed: "fast", difficult: true },
  [ManeuverConstants.Fast_then_right]: { direction: "right", speed: "fast", difficult: false },
  [ManeuverConstants.Right_then_fast]: { direction: "right", speed: "fast", difficult: false },
};

export function getManeuverDetails(maneuver: string): ManeuverDetails | null {
  return ManeuverDetailsMap[maneuver] || null;
}

export function selectManeuvers(pred: (details: ManeuverDetails) => boolean): string[] {
  return Object.keys(ManeuverDetailsMap).filter(maneuver => pred(ManeuverDetailsMap[maneuver]));
}